listenbrainz_spark/stats/user.py

```diff
@@ -0,0 +1,214 @@
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'listen_count': row.cnt,
+        })
+    print("Query to calculate recording stats processed in %.2f s" % (time.time() - t0))
+    return recordings
+
+def get_releases(table):
+    """
+    Get releases information (release_name, release_mbid etc) for every user
+    ordered by listen count (number of times a user has listened to tracks 
+    which belong to a particular release).
+
+    Args:
+        table: name of the temporary table
+
+    Returns:
+        artists: A dict of dicts which can be depicted as:
+                {'user1' : {'release_name' : 'xxxx', 'release_msid' : 'xxxx',
+                    'release_mbid' : 'xxxx', 'artist_name' : 'xxxx', 'artist_msdid' : 
+                    'xxxx', 'artist_mbids' : 'xxxx', count' : 'xxxx'}, 'user2' : {...}}
+    """
+    t0 = time.time()
+    query = run_query("""
+            SELECT user_name
+                 , release_name
+                 , release_msid
+                 , release_mbid
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(release_msid) as cnt
+              FROM %s
+          GROUP BY user_name, release_name, release_msid, release_mbid, artist_name, artist_msid, artist_mbids
+          ORDER BY cnt DESC
+        """ % (table))
+    rows = query.collect()
+    releases = defaultdict(list)
+    for row in rows:
+        releases[row.user_name].append({
+            'release_name': row.release_name,
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'artist_name': row.artist_name,
+            'artist_msid': row.artist_msid,
+            'artist_mbids': row.artist_mbids,
+            'listen_count': row.cnt,
+        })
+    print("Query to calculate release stats processed in %.2f s" % (time.time() - t0))
+    return releases
+
+def main(app_name):
+    """
+    Calculate statistics of users for previous month.
+
+    Args:
+        app_name: Application name to uniquely identify the submitted
+                  application amongst other applications
+
+    """
+    t0 = time.time()
+    try:
+        listenbrainz_spark.init_spark_session(app_name)
+    except Exception as err:
+        logging.error("Cannot initialize spark session: %s / %s. Aborting." % (type(err).__name__, str(err)))
+        sys.exit(-1)
+    date = date_for_stats_calculation()
+    try:
+        df = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'.format(config.HDFS_CLUSTER_URI, date.year, date.month))
+        print("Loading dataframe...")
+    except Exception as err:
+        logging.error("Cannot read files from HDFS: %s / %s. Aborting." % (type(err).__name__, str(err)))
+        sys.exit(-1)
+    df.printSchema()
+
+    table = 'listens_{}'.format(datetime.strftime(date, '%Y_%m'))
+    print(table)
+    df.registerTempTable(table)
```

Comment:
Could this fail? If so, it should be in a try block.

Reply:
Not really, but a general try block would be good, I guess.